### ISSUE TITLE
HTTP APIs Components

### ISSUE THREAD

**Issue by eahefnawy** [serverless/components#432]
As we start writing complex APIs for different use cases, we need to take a big picture overview of what components we'll need, and how they fit together. This issue proposes a certain architecture that attempts to cover as many uses cases as possible.

## api
- **description:** creates an api from scratch. Can optionally take a swagger or OpenApi template.
- **Usecase:** for complete flexablity, like if you want to deploy an API that uses existing functions and/or roles.
- **children:** none

```yml
api:
  component: "@serverless/api"
  inputs:
    name: my-api
    template: ./swagger.json # optional
``` 

## restApi
- **description:** creates an api from scratch along with defined endpoints, methods & function code.
- **Usecase:** if you want to easily and quickly spin up a rest api backend.
- **children:** api (using swagger), function, role, s3

```yml
restApi:
  component: "@serverelss/components"
  inputs:
    name: my-rest-api
    routes:
      /users:
        get: ./code/users/get # code dir
```

## mono
- **description:** creates and manages an entire api and proxy all endpoints & methods to a single function
- **Usecase:** if you're writing an api from scratch and you want to use a single lambda
- **children:** api, function, role, s3

```yml
mono:
  component: "@serverless/mono"
  inputs:
    name: my-app
    code: ./
```  

## endpoint
- **description:** creates and manages an endpoint for the provided api along with their methods & code
- **Usecase:** if you want to create a microservice or an endpoint to an existing api
- **children:** function, role

```yml
endpoint:
  component: "@serverless/endpoint"
  inputs:
    api: ${comp:api.id}
    path: /users
    methods:
      get: ./code/users/get
```   

## microservice
- **description:** creates and manages a microservice in an existing api along with its endpoints, methods & function code
- **Usecase:** if you want to create a microservice to an existing api
- **children:** function, role

```yml
microservice:
  component: "@serverless/microservice"
  inputs:
    api: ${comp:api.id}
    routes:
      /users:
        get: ./code/users/get
```

## rpc
- **description:** creates a single endpoint in an existing api with all methods routed to a single function
- **Usecase:** if you want to create an RPC api (rather than rest api)
- **children:** function, role

```yml
rpc:
  component: "@serverless/rpc"
  inputs:
    api: ${comp:api.id}
    name: sendSms
```   

Open for feedback...

**Comment by eahefnawy** [comment]
Also, as we write higher level components (ie. Images, Users...etc), which one of those patterns should we choose?

I guess it depends on the component, but for the most part, I think the microservice component is the most "unopinionated" one and fits pretty well.

**Comment by pmuens** [comment]
Thanks for putting this together @eahefnawy 👍

Here are some thoughts after reading through this:

**Consolidating `api` and `restApi`**

It seems like the only real difference between `api` and `restApi` is that the former takes a swagger template and the latter takes routes. Given that most of the time APIs don't adhere to `REST` we could consolidate both into the `api` component which takes either a `swagger` template or `routes`.

**`mono` component**

Isn't it possible to build this with the `api` component by leveraging `ANY`? Why do we need an extra component for this?

**`endpoint` component**

I like the idea of it but it's unclear for me how this can be used in production since overwrites are inevitable.

I feel like most of the time you want multiple endpoints (could also be wrong). If that's the case you'd need multiple endpoint components. We could extend the `api` component to take an `apiId` which is then used to wrap en existing API. The `routes` defined on the `api` component are then merged into the existing API. This way there's no need for multiple `endpoint` components.

**`microservices` component**

It seems like this is what I've described above. The functionality here could be added to the `api` component.

**`rpc` component**

I don't have enough information about this. Not sure how this would be used...

---

In summary I'd vote for a fat `api` component which provides all the features which can be found in the separate components. This way there's only one component you need to know about and we could still split the monolith up into separate components as we see patterns emerge.

**Comment by eahefnawy** [comment]
Thanks for your feedback @pmuens ... I guess you're right about api vs restApi. However one major difference between the api component and the others is that the others take code as input and use the function component as a child. So you can't really use the api component to replace the others without having to use other components along with it.

**Comment by ac360** [comment]
Great write-up @eahefnawy.

* `api` and `restApi` should definitely be merged imo.
* `mono` makes sense because it combines code + http communication into a specific runtime experience.  It goes above and beyond endpoints.
* `microservice` I think should just be `function`.  It's more relevant in the serverless domain.
* `rpc` I haven't heard anyone ask about this.  Don't think we should focus on this now.
* `endpoint` Yes, unless there is another way to solve the problem.

As always, component design should be focused on priority use-cases.  IMO, this is the priority use-case:

* Creating a large REST API and enabling developers to work on individual endpoints/services separately/autonomously without affecting other aspects of the REST API or getting blocked by whomever controls it.

My hypothesis on the ideal MVP is:

**Component 1**
```yaml
name: webapp-api

rest-api:
  component: @serverless/aws-rest-api
  inputs:
    domain:  webapp.com
    region: us-east-1

exports:
  restApiId:  ${comp:rest-api.id}
```

**Component 2**
```yaml
name: users-api

createUser:
  component: @serverless/function
  inputs:
    code: ./index.create
    route:
      path: /users/create
      method: post
      apiId:  ${comp:webapp-api.restApiId}
```

The Function in Component 2 SHOULD be designed so that developers can work on this Function/Microservice attached to a larger REST API autonomously.

The Function in Component 2 SHOULD be able to create/update/delete a SINGLE endpoint on an existing REST API IF given an ID.

The Function in Component 2 SHOULD NOT create the endpoint if it already exists and WAS NOT created by this Function.  It SHOULD throw an error saying the path is taken.  It's state SHOULD keep track of the endpoint it has created.

The Function in Component 2 SHOULD be able to update/delete the endpoint IF it created the endpoint.

The Function in Component 2 SHOULD remove it's existing endpoint if the endpoint configuration has been updated, before provisioning a new one.

This pattern SHOULD work across many other Components containing Functions, all sharing the same REST API created in Component 1.

**Comment by eahefnawy** [comment]
That sounds good @ac360 ... It seems that the Function component would take path/method/apiId as inputs. That's exactly how I thought about the RPC component. I do agree about calling it function  for the serverless domain though 👍 ... as long as this use case is covered.

I think there's a misconception on what a microservice is though. I thought that a microservice would likely have multiple endpoints/methods (ie. users microservice would likely have multiple functions). So I'm not sure that Function = Microservice.

With that said though, if we do have a Function component as you described, we will be able to build microservices on top of that using multiple instances of the Function component. I guess that's what you're going for anyway. I was thinking about putting a layer in the middle called "microservice" that a users component could just use as a single instance to create a microservice. But that's probably too much abstraction.

**Comment by eahefnawy** [comment]
My only worry regarding the Function component is that it's now http focused. So we're still not sure about how it would handle other events. But I guess that's another future discussion.

**Side Note:** Maybe the function/events pattern we're using in v1 is still a good idea for components. In that case http would be an event along with others. But I'm gonna leave that open for now...

**Comment by eahefnawy** [comment]
FYI I've opened two seperate issues as action items resulting from this discussion:

- #438 
- #440

**Comment by ac360** [comment]
I believe the terms usually are defined this way:

microservice - 1 task, 1 endpoint.
service - multiple endpoints and tasks related to a common business domain.
monolith - an entire app.
